src/main/java/org/opensearch/security/user/UserService.java

```diff
@@ -0,0 +1,181 @@
+            throw new UserServiceException(RESTRICTED_CHARACTER_USE_MESSAGE + restrictedContents);
+        }
+
+        // if password is set, it takes precedence over hash
+        final String plainTextPassword = securityJsonNode.get("password").asString();
+        final String origHash = securityJsonNode.get("hash").asString();
+        if (plainTextPassword != null && plainTextPassword.length() > 0) {
+            contentAsNode.remove("password");
+            contentAsNode.put("hash", hash(plainTextPassword.toCharArray()));
+        } else if (origHash != null && origHash.length() > 0) {
+            contentAsNode.remove("password");
+        } else if (plainTextPassword != null && plainTextPassword.isEmpty() && origHash == null) {
+            contentAsNode.remove("password");
+        }
+
+        final boolean userExisted = internalUsersConfiguration.exists(accountName);
+
+        // sanity checks, hash is mandatory for newly created users
+        if (!userExisted && securityJsonNode.get("hash").asString() == null) {
+            throw new UserServiceException(NO_PASSWORD_OR_HASH_MESSAGE);
+        }
+
+        // for existing users, hash is optional
+        if (userExisted && securityJsonNode.get("hash").asString() == null) {
+            // sanity check, this should usually not happen
+            final String hash = ((Hashed) internalUsersConfiguration.getCEntry(accountName)).getHash();
+            if (hash == null || hash.length() == 0) {
+                throw new UserServiceException("Existing user " + accountName + " has no password, and no new password or hash was specified.");
+            }
+            contentAsNode.put("hash", hash);
+        }
+
+        internalUsersConfiguration.remove(accountName);
+        contentAsNode.remove("name");
+
+        internalUsersConfiguration.putCObject(accountName, DefaultObjectMapper.readTree(contentAsNode,  internalUsersConfiguration.getImplementingClass()));
+        return internalUsersConfiguration;
+    }
+
+    private void verifyServiceAccount(SecurityJsonNode securityJsonNode, String accountName) {
+
+        final String plainTextPassword = securityJsonNode.get("password").asString();
+        final String origHash = securityJsonNode.get("hash").asString();
+
+        if (plainTextPassword != null && plainTextPassword.length() > 0) {
+            throw new UserServiceException(SERVICE_ACCOUNT_PASSWORD_MESSAGE + accountName);
+        }
+
+        if (origHash != null && origHash.length() > 0) {
+            throw new UserServiceException(SERVICE_ACCOUNT_HASH_MESSAGE + accountName);
+        }
+    }
+
+    /**
+     * This will be swapped in for a real solution once one is decided on.
+     *
+     * @return A password for a service account.
+     */
+    private String generatePassword() {
+        String generatedPassword = "superSecurePassword";
+        return generatedPassword;
+    }
+}
```

Comment:
👏🏼 Great job moving in all validations and user related service methods into this new class. Better for readability!